src/dex/woo-fi/woo-fi.ts

```diff
@@ -0,0 +1,556 @@
+          target: this.config.woOracleAddress,
+          callData: this.wooIfaces.oracle.encodeFunctionData('infos', [
+            t.address,
+          ]),
+        },
+        {
+          target: this.config.wooPPAddress,
+          callData: this.wooIfaces.PP.encodeFunctionData('tokenInfo', [
+            t.address,
+          ]),
+        },
+      ]),
+    );
+
+    calldata.push({
+      target: this.config.wooPPAddress,
+      callData: this.wooIfaces.PP.encodeFunctionData('tokenInfo', [
+        this.quoteTokenAddress,
+      ]),
+    });
+
+    const data = await this.dexHelper.multiContract.methods
+      .aggregate(calldata)
+      .call({}, blockNumber || 'latest');
+
+    // Last request is standalone
+    const maxNumber = calldata.length - 1;
+
+    const [baseFeeRates, baseInfos, baseTokenInfos, quoteTokenInfo] = [
+      // Skip two as they are infos abd tokenInfo
+      _.range(0, maxNumber, 3).map(index =>
+        this.wooIfaces.fee.decodeFunctionResult(
+          'feeRate',
+          data.returnData[index],
+        ),
+      ),
+      _.range(1, maxNumber, 3).map(index =>
+        this.wooIfaces.oracle.decodeFunctionResult(
+          'infos',
+          data.returnData[index],
+        ),
+      ),
+      _.range(2, maxNumber, 3).map(index =>
+        this.wooIfaces.PP.decodeFunctionResult(
+          'tokenInfo',
+          data.returnData[index],
+        ),
+      ),
+      this.wooIfaces.PP.decodeFunctionResult(
+        'tokenInfo',
+        data.returnData[maxNumber],
+      ),
+    ];
+
+    const state: PoolState = {
+      feeRates: {},
+      tokenInfos: {},
+      tokenStates: {},
+    };
+    this._fillTokenInfoState(state, this.quoteTokenAddress, quoteTokenInfo);
+
+    baseFeeRates.map((value, index) => {
+      const tokenAddress = this.baseTokens[index].address.toLowerCase();
+      state.feeRates[tokenAddress] = BigInt(value[0]._hex);
+    });
+    baseInfos.map((value, index) => {
+      const tokenAddress = this.baseTokens[index].address.toLowerCase();
+      state.tokenStates[tokenAddress] = {
+        priceNow: BigInt(value.price._hex),
+        coeffNow: BigInt(value.coeff._hex),
+        spreadNow: BigInt(value.spread._hex),
+      };
+    });
+    baseTokenInfos.map((values, index) => {
+      this._fillTokenInfoState(state, this.baseTokens[index].address, values);
+    });
+
+    return state;
+  }
+
+  async initializePricing(blockNumber: number) {
+    this.latestState = await this.fetchStateForBlockNumber(blockNumber);
+    this.latestBlockNumber = blockNumber;
+
+    this.tokenByAddress = {};
+
+    // Normalising to toLowerCase()
+    this.config.quoteToken.address = this.quoteTokenAddress;
+    this.tokenByAddress[this.quoteTokenAddress] = this.config.quoteToken;
+
+    for (const baseToken of this.baseTokens) {
+      const baseTokenAddress = baseToken.address.toLowerCase();
+      baseToken.address = baseTokenAddress;
+      this.tokenByAddress[baseTokenAddress] = baseToken;
+    }
```

Comment:
this should be added to constructor! tokenByAddress should always be defined after the instance is created

Reply:
Fixed